Errors and exceptions handling improvement

MrXANA91 opened the pull request:
This PR aims to fix #36 and to generally improve the exceptions potentially raised by the integration using the Home Assistant exception classes

If the Switchbot servers return a 500 status, the integration will send a warning to the Home Assistant logs.
It will wait 0.5s and will then try to resend the same command.
After 5 consecutive 500 status failures, it will (should ?) abort and raise an error to Home Assistant.

MrXANA91 committed:
Migrating from default Python exception classes to HomeAssistant exception classes

MrXANA91 committed:
Raising special exception for 500 return status

MrXANA91 committed:
Added a filtered request to try to filter out the 500 returned status

MrXANA91 committed:
Fixed a bug using time sleep

MrXANA91 commented:
Currently testing this version on my HA instance.

I have tested by forcing a returned status to 500 at random, and it effectively resends the same request after 0.5s repeatedly, until the return status is 200. If the 5 tries all failed with 500, then it aborts and raise a `SwitchbotInternal500Error` exception to the Home Assistant core.

MrXANA91 committed:
Tweaked logs when catching a SwitchbotInternal500Error

MrXANA91 committed:
Merge branch 'KiraPC:main' into 500SwitchbotServersFix

MrXANA91 commented:
And of course, since I started testing this fix, the 500 error status no longer appear for me ...

Making it ready for review anyway, still testing

KiraPC commented on custom_components/switchbotremote/client/client.py:
```diff
@@ -46,24 +51,48 @@ def request(self, method: str, path: str, **kwargs) -> Any:
         _LOGGER.debug(f"Call service {url} OK")
         return response_in_json
+    
+    def filtered_request(self, method: str, path: str, **kwargs) -> Any:
```
@MrXANA91 I don't like to have unused code, what about the request api? Is it still used? 

In case not, change request api instead of add a new one, and add a parameter for "retry" that could be MAX_TRRIES as default and with 0 you should have the same behavior of the old request implementation.

MrXANA91 replied:
If you check line 73, you can see a call to the request API. `filtered_request` just wraps the current `request` method to catch any 500 status that could potentially appear.

First request is sent (considered first try). If a 500 status is caught and `MAX_TRIES` is greater than 1, the request is resent after waiting 500ms (or the value of `DELAY_BETWEEN_TRIES_MS`). If every requests throw a 500 error code, a total number of `MAX_TRIES` (here is 5) requests will be sent.

If you set the value of `MAX_TRIES` to 1, this code will behave exactly like it currently does in the release.

KiraPC replied:
mmmm ok, good point. Anyway, I don't like the name, why it is filtered? 

Could you use for:else (https://www.w3schools.com/python/gloss_python_for_else.asp) returning the result instead of breaking the loop?

KiraPC replied:
Anyway, I prefer to make request "private" -> so you could name your method request and I would like to accept the retry as parameter that per default it is equals to MAX_RETRIES

MrXANA91 replied:
Got it, I'll make the changes.

Do you want the delay value as a parameter as well?

KiraPC replied:
Yup, that's great.

Thanks a lot for your help and passion!

KiraPC reviewed (changes requested):
Just a little feedback. Other is ok

MrXANA91 committed:
Changed requests methods names
- setting original 'request' method as private
- renamed 'filtered_request' to 'request'

MrXANA91 committed:
Added parameters to new request method
- maxNumberOfTrails defaults to MAX_TRIES
- delayMSBetweenTrials defaults to DELAY_BETWEEN_TRIES_MS

MrXANA91 committed:
Improved clarity + fixed formatted debug and error messages

MrXANA91 commented on custom_components/switchbotremote/client/client.py:
```diff
@@ -39,22 +44,43 @@ def headers(self):
+        for tryNumber in range(maxNumberOfTrials):
+            try:
+                result = self.__request(method, path, **kwargs)
+                break
```
I have to keep the `break` statement there in order to get out of the loop if the request succeeds.
I added the `else` statement which is better than checking the `tryNumber` but I don't think that I can get rid of the `break` statement unless I completely rethink the way the trials are handled.

KiraPC replied:
@MrXANA91 As I suggested, you should use a return statement in case of success directly inside the try block, and the else statement is executed just when the loop end without result

MrXANA91 committed:
Replacing 'break' statement with the 'return' statement

KiraPC commented on custom_components/switchbotremote/client/client.py:
```diff
@@ -39,22 +44,41 @@ def headers(self):
+        """Try to send the request.
+        If the server returns a 500 Internal error status, will retry until it succeeds or it passes a threshold of max number of tries.
+        Any other error will be thrown."""
+        for tryNumber in range(maxNumberOfTrials):
```
@MrXANA91 wheere do you increase the try number?

KiraPC reviewed (approved):
Have you tested and everything works as expected?

MrXANA91 commented:
Going back to draft during testing

Not tested yet, I'll let you know

MrXANA91 commented:
A few minutes ago, when trying to send a `turn_off` command, i've noticed a unusual slightly longer delay between me pressing the button and the AC signaling a received command.
When checking the logs, here is what I've found :

![1-Caught 500](https://github.com/KiraPC/ha-switchbot-remote/assets/12546483/24ebfcd0-bbf1-4ed5-aa49-e995692a279b)

```
...
2024-04-08 19:35:00.318 DEBUG (SyncWorker_27) [custom_components.switchbotremote.client.remote] Sending command turnOff
2024-04-08 19:35:00.318 DEBUG (SyncWorker_27) [custom_components.switchbotremote.client.remote] Command payload {'commandType': 'command', 'command': 'turnOff', 'parameter': 'default'}
2024-04-08 19:35:00.318 DEBUG (SyncWorker_27) [custom_components.switchbotremote.client.client] Calling service https://api.switch-bot.com/v1.1/devices/***/commands
2024-04-08 19:35:00.662 DEBUG (SyncWorker_27) [custom_components.switchbotremote.client.client] Received http error 500 {"message": "Internal server error"}
2024-04-08 19:35:00.662 WARNING (SyncWorker_27) [custom_components.switchbotremote.client.client] Caught returned status 500 from SwitchBot API server
2024-04-08 19:35:00.662 DEBUG (SyncWorker_27) [custom_components.switchbotremote.client.client] tryNumber = 0, waiting 500 ms
2024-04-08 19:35:01.164 DEBUG (SyncWorker_27) [custom_components.switchbotremote.client.client] Calling service https://api.switch-bot.com/v1.1/devices/***/commands
2024-04-08 19:35:01.694 DEBUG (SyncWorker_27) [custom_components.switchbotremote.client.client] Call service https://api.switch-bot.com/v1.1/devices/***/commands OK
...
```

So the fix is actually working properly. Marking it as ready!

KiraPC commented:
Great, merging
